File: node/verifyToken.js
const crypto = require('crypto');
require('dotenv').config({ path: '../.env' });

const SECRET_KEY = process.env.SECRET_KEY;

const verifyToken = (token) =>{
    if (!token || typeof token !== 'string') return null;
     
    // Ensure token has 3 parts only
    const temp = token.split('.');
    if (temp.length !== 3) return null;

    const unsignedToken = `${temp[0]}.${temp[1]}`;

    const signature = crypto
        .createHmac('sha256', SECRET_KEY)
        .update(unsignedToken)
        .digest('base64url'); 
    
    // timingSafeEqual compares raw binary data
    const signatureBuffer = Buffer.from(signature);
    const providedSignatureBuffer = Buffer.from(temp[2]);

    // timingSafeEqual requires buffers of the exact same length
    if (signatureBuffer.length !== providedSignatureBuffer.length) {
        return null
    }

    const isValid = crypto.timingSafeEqual(signatureBuffer, providedSignatureBuffer);
    return isValid ? temp[1] : null;
}
module.exports = verifyToken;

File: go/verifyToken.go
package main

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/base64"
	"log"
	"os"
	"strings"
)

func VerifyToken(token string) bool {
	if token == "" {
		return false
	}
	temp := strings.Split(token, ".")
	if len(temp) != 3 {
		return false
	}
	unsignedToken := strings.Join(temp[:2], ".")
	log.Println(unsignedToken)

	// Default secret key or pull from environment
	secretKey := os.Getenv("SECRET_KEY")

	// 1. Create HMAC SHA256 generator
	mac := hmac.New(sha256.New, []byte(secretKey))

	// 2. Hash the unsigned token (equivalent to .update())
	mac.Write([]byte(unsignedToken))
	expectedMAC := mac.Sum(nil)

	// 3. Encode to Base64URL without padding (equivalent to .digest('base64url'))
	expectedSignature := base64.RawURLEncoding.EncodeToString(expectedMAC)

	// 4. constant-time comparison (equivalent to timingSafeEqual)
	return hmac.Equal([]byte(expectedSignature), []byte(temp[2]))

}
